cmd/mistryd/worker.go

```diff
@@ -98,10 +97,40 @@ func (s *Server) Work(ctx context.Context, j *Job) (buildInfo *types.BuildInfo,
 	}
 
 	log.Printf("Creating new build directory...")
-	cleanupPending, err := j.BootstrapBuildDir(s.cfg.FileSystem, log)
+	err = j.BootstrapBuildDir(s.cfg.FileSystem, log)
+	if err != nil {
+		err = workErr("could not bootstrap build dir", err)
+	}
 
+	// moves the pending directory to the ready one
 	defer func() {
-		if cleanupPending {
+		rerr := os.Rename(j.PendingBuildPath, j.ReadyBuildPath)
+		if rerr != nil {
+			errstr := "could not move pending path"
+			if err == nil {
+				err = fmt.Errorf("%s; %s", errstr, rerr)
+			} else {
+				err = fmt.Errorf("%s; %s | %s", errstr, rerr, err)
+			}
+		}
+
+		// When there are no errors the LatestBuildPath can be updated
+		// to point to the last build. Otherwise it remains unchanged.
+		if err == nil {
+			_, err = os.Lstat(j.LatestBuildPath)
+			if err == nil {
+				err = os.Remove(j.LatestBuildPath)
+				if err != nil {
+					err = workErr("could not remove latest build link", err)
+					return
+				}
+			}
+			err = os.Symlink(j.ReadyBuildPath, j.LatestBuildPath)
+			if err != nil {
+				err = workErr("could not create latest build link", err)
+			}
+			// An error at this point indicates that there is a hanging pending path.
+		} else {
 			derr := s.cfg.FileSystem.Remove(j.PendingBuildPath)
```

Comment:
We don't need to `Remove(j.PendingBuildPath)` any more. We  always `os.Rename(j.PendingBuildPath, j.ReadyBuildPath)`. If that move fails, this means that we either have a bug or there is some system/permission error and manual intervention is needed by an admin to solve the issue. So I think we should never do  `Remove(j.PendingBuildPath)` here any more

Reply:
True.